### PR TITLE
implementation : divide replicas by static weight evenly

### PR DESCRIPTION
**Pull request by chaosi-zju**
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

divide replicas by static weight evenly

code implementation of proposal #4218

**Which issue(s) this PR fixes**:

Fixes part of #4220

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

File: pkg/util/helper/binding.go [hunk 1]
```diff
@@ -45,19 +48,12 @@ func (p ClusterWeightInfoList) Less(i, j int) bool {
 	if p[i].Weight != p[j].Weight {
 		return p[i].Weight > p[j].Weight
 	}
-	return p[i].ClusterName < p[j].ClusterName
-}
-
-// SortClusterByWeight sort clusters by the weight
-func SortClusterByWeight(m map[string]int64) ClusterWeightInfoList {
```
Comment (chaosi-zju) [inline pkg/util/helper/binding.go:52]: This function is never used, just only used in one test case, so I moved it to UT code.

File: pkg/util/helper/binding.go [hunk 2]
```diff
@@ -125,21 +121,29 @@ func (a *Dispenser) TakeByWeight(w ClusterWeightInfoList) {
...
+			ClusterName:  targetCluster.Name,
+			Weight:       int64(targetCluster.Replicas),
+			LastReplicas: lastReplicas,
 		})
 	}
 	return weightList
 }
 
 // SpreadReplicasByTargetClusters divides replicas by the weight of a target cluster list.
-func SpreadReplicasByTargetClusters(numReplicas int32, tcs, init []workv1alpha2.TargetCluster) []workv1alpha2.TargetCluster {
```
Comment (chaosi-zju) [inline pkg/util/helper/binding.go:140]: tcs and init is confusing, so I just renamed it

File: pkg/util/helper/binding.go [hunk 3]
```diff
@@ -45,19 +48,12 @@ func (p ClusterWeightInfoList) Less(i, j int) bool {
...
-
-// SortClusterByWeight sort clusters by the weight
-func SortClusterByWeight(m map[string]int64) ClusterWeightInfoList {
-	p := make(ClusterWeightInfoList, len(m))
-	i := 0
-	for k, v := range m {
-		p[i] = ClusterWeightInfo{k, v}
-		i++
+	if p[i].LastReplicas != p[j].LastReplicas {
+		return p[i].LastReplicas > p[j].LastReplicas
```
Comment (RainbowMango) [inline pkg/util/helper/binding.go:52]: Please add a comment to explain why need to compare `last replicas`.

File: pkg/util/helper/binding.go [hunk 4]
```diff
@@ -45,19 +48,12 @@ func (p ClusterWeightInfoList) Less(i, j int) bool {
...
 	}
-	sort.Sort(p)
-	return p
+	const maxRandomNum = 100
+	randomNum, err := rand.Int(rand.Reader, big.NewInt(maxRandomNum)) // [0, 100)
+	return err == nil && randomNum.Cmp(big.NewInt(maxRandomNum/2)) >= 0
```
Comment (RainbowMango) [inline pkg/util/helper/binding.go:56]: Add comments to explain what this algorithm does.

File: pkg/util/helper/binding.go (hunk 2 above)
Comment (RainbowMango) [reply pkg/util/helper/binding.go:140]: I don't think so, `tcs` means `target clusters`, it's not the same meaning of `available`.

File: pkg/util/helper/binding.go [hunk 5]
```diff
@@ -125,21 +121,29 @@ func (a *Dispenser) TakeByWeight(w ClusterWeightInfoList) {
 }
 
 // GetStaticWeightInfoListByTargetClusters constructs a weight list by target cluster slice.
-func GetStaticWeightInfoListByTargetClusters(tcs []workv1alpha2.TargetCluster) ClusterWeightInfoList {
-	weightList := make(ClusterWeightInfoList, 0, len(tcs))
-	for _, result := range tcs {
+func GetStaticWeightInfoListByTargetClusters(available, scheduled []workv1alpha2.TargetCluster) ClusterWeightInfoList {
```
Comment (RainbowMango) [inline pkg/util/helper/binding.go:124]: My suggestion is not to rename the parameters, as `tcs` is used everywhere, if you think it is worth renaming it, do it by a separate PR and add some comments on it.

File: pkg/util/helper/binding.go (hunk 3 above)
Comment (chaosi-zju) [reply pkg/util/helper/binding.go:55]: done

File: pkg/util/helper/binding.go (hunk 2 above)
Comment (chaosi-zju) [reply pkg/util/helper/binding.go:140]: done

File: pkg/util/helper/binding.go (hunk 5 above)
Comment (chaosi-zju) [reply pkg/util/helper/binding.go]: done

File: pkg/util/helper/binding.go [hunk 6]
```diff
@@ -45,19 +48,18 @@ func (p ClusterWeightInfoList) Less(i, j int) bool {
...
+	// so in order to keep the inertia in this scheduling, such clusters should also be prioritized
+	if p[i].LastReplicas != p[j].LastReplicas {
+		return p[i].LastReplicas > p[j].LastReplicas
+	}
+	// when last scheduling replicas is also equal, sort by random,
+	// first generate a random number within [0, 100) range,
+	// then return < if the actual number is in [0, 50) range, return > if is in [50, 100) range
+	const maxRandomNum = 100
+	randomNum, err := rand.Int(rand.Reader, big.NewInt(maxRandomNum))
+	return err == nil && randomNum.Cmp(big.NewInt(maxRandomNum/2)) >= 0
```
Comment (RainbowMango) [inline pkg/util/helper/binding.go:62]: What's the difference between the random algorithm with `math/rand`? 
```golang
	mathrand.New(mathrand.NewSource(time.Now().UnixNano()))
	randi := mathrand.Int()
	randj := mathrand.Int()
	return randi > randj
```

File: pkg/util/helper/binding.go [hunk 7]
```diff
@@ -125,20 +127,28 @@ func (a *Dispenser) TakeByWeight(w ClusterWeightInfoList) {
...
+			LastReplicas: lastReplicas,
 		})
 	}
 	return weightList
 }
 
 // SpreadReplicasByTargetClusters divides replicas by the weight of a target cluster list.
 func SpreadReplicasByTargetClusters(numReplicas int32, tcs, init []workv1alpha2.TargetCluster) []workv1alpha2.TargetCluster {
-	weightList := GetStaticWeightInfoListByTargetClusters(tcs)
+	weightList := GetStaticWeightInfoListByTargetClusters(tcs, init)
```
Comment (RainbowMango) [inline pkg/util/helper/binding.go:151]: The `SpreadReplicasByTargetClusters` is used by `DynamicWeight` case. 
Do you mean we have to keep the assignment stable?

File: pkg/scheduler/core/assignment_test.go [hunk 8]
```diff
@@ -591,8 +591,11 @@ func Test_dynamicScaleUp(t *testing.T) {
 		candidates []*clusterv1alpha1.Cluster
 		object     *workv1alpha2.ResourceBindingSpec
 		placement  *policyv1alpha1.Placement
-		want       []workv1alpha2.TargetCluster
-		wantErr    bool
+		// want specifies exactly one detail desired result (lose effect when wants defined)
+		want []workv1alpha2.TargetCluster
+		// wants specifies multi possible desired result, any one got is expected
+		wants   [][]workv1alpha2.TargetCluster
```
Comment (RainbowMango) [inline pkg/scheduler/core/assignment_test.go:597]: Can we deprecate the `want` and put all expectations into `wants`?

File: pkg/util/helper/binding_test.go [hunk 9]
```diff
@@ -41,7 +42,10 @@ func TestDispenseReplicasByTargetClusters(t *testing.T) {
 	tests := []struct {
 		name string
 		args args
+		// want specifies exactly one detail desired result (lose effect when wants defined)
 		want []workv1alpha2.TargetCluster
+		// wants specifies multi possible desired result, any one got is expected
+		wants [][]workv1alpha2.TargetCluster
```
Comment (RainbowMango) [inline pkg/util/helper/binding_test.go:48]: ditto.

File: pkg/scheduler/core/assignment_test.go (hunk 8 above)
Comment (chaosi-zju) [reply pkg/scheduler/core/assignment_test.go:597]: This is also possible. I initially didn't want to affect too many lines

File: pkg/util/helper/binding.go (hunk 7 above)
Comment (chaosi-zju) [reply pkg/util/helper/binding.go:151]: yes, from our previous implementation, DynamicWeight is just a special StaticWeight

File: pkg/util/helper/binding.go (hunk 6 above)
Comment (chaosi-zju) [reply pkg/util/helper/binding.go:62]: There's not much difference, just to achieve a randomness

Comment (chaosi-zju) [reply pkg/util/helper/binding.go:62]: shall I change to this method ?

Comment (chaosi-zju) [reply pkg/util/helper/binding.go:62]: shall I change to your approch ? @RainbowMango

Comment (RainbowMango) [reply pkg/util/helper/binding.go:62]: Yeah, I get it, thanks for the reminder. `match/rand` package might bring unnecessary concerns, especially security scanners.

**Review by RainbowMango — COMMENTED** [review]
/lgtm

Wait for a while to see if others have any concern.

**Review by RainbowMango — APPROVED** [review]
/approve
